PopPUNK/network.py

```diff
@@ -167,83 +194,124 @@ def extractReferences(G, dbOrder, outPrefix, existingRefs = None, threads = 1):
         index_lookup = {v:k for k,v in enumerate(dbOrder)}
         reference_indices = set([index_lookup[r] for r in references])
 
-    # Each component is independent, so can be multithreaded
-    components = gt.label_components(G)[0].a
-
-    # Turn gt threading off and on again either side of the parallel loop
-    if gt.openmp_enabled():
-        gt.openmp_set_num_threads(1)
-
-    # Cliques are pruned, taking one reference from each, until none remain
-    with Pool(processes=threads) as pool:
-        ref_lists = pool.map(partial(cliquePrune,
-                                        graph=G,
-                                        reference_indices=reference_indices,
-                                        components_list=components),
-                             set(components))
-    # Returns nested lists, which need to be flattened
-    reference_indices = set([entry for sublist in ref_lists for entry in sublist])
+    if use_gpu:
+
+        # load CUDA libraries
+        try:
+            import cugraph
+            import cudf
+        except ImportError as e:
+            sys.stderr.write("cugraph and cudf unavailable\n")
+            raise ImportError(e)
+    
+        # For large network, use more approximate method for extracting references
+        reference = {}
+        # Leiden method has resolution parameter - higher values give greater precision
+        component_assignments, score = cugraph.leiden(G, resolution = 1.0)
+        # group by partition, which becomes the first column, so retrieve second column
+        reference_index_df = component_assignments.groupby('partition').nth(0)
+        reference_indices = reference_index_df['vertex'].to_arrow().to_pylist()
+        
+        # Order found references as in mash sketch files
+        reference_names = [dbOrder[int(x)] for x in sorted(reference_indices)]
+        refFileName = writeReferences(reference_names, outPrefix)
+        
+        # Extract reference edges
+        G_df = G.view_edge_list()
+        G_df.columns = ['source','destination']
+        G_ref_df = G_df[G_df['source'].isin(reference_names) & G_df['destination'].isin(reference_names)]
+        # Add self-loop if needed
+        max_in_df = np.amax([G_df['source'].max(),G_df['destination'].max()])
+        max_in_vertex_labels = len(reference_names)-1
+        if max_in_df.item() != max_in_vertex_labels:
+            G_self_loop = cudf.DataFrame()
+            G_self_loop['source'] = [max_in_vertex_labels]
+            G_self_loop['destination'] = [max_in_vertex_labels]
+            G_ref_df = cudf.concat([G_ref_df,G_self_loop], ignore_index = True)
```

Comment:
```suggestion
            G_ref_df = cudf.concat([G_ref_df, G_self_loop], ignore_index = True)
```

Reply:
Do these self loops cause any issues later on? Perhaps we should remove all self loops when the cugraph steps are done?